infrastructure/containerd/image_service_integration_test.go

```diff
@@ -0,0 +1,291 @@
+//
+// == Chapter I: Pull and mount
+//
+// Using ImageService, we tries to Pull an image.
+//
+// We don't have separate Mount function (yet), we try to PullAndMount, with
+// this, we can be sure ImageService can pull and image from a repository and
+// can mount it.
+//
+// To avoid false positive result, we try to PullAndMount an image, we know
+// it's definitely not there, and we expect it to be failed.  If it does not
+// fail, something fishy is swimming under the hood.
+//
+// After PullAndMount, the real image should be mounted, but the fake one
+// shouldn't be. At the end we should have only one image available locally.
+//
+// == Chapter II: Snapshots
+//
+// As we already pull and mounted an image, we expect one snapshot to be there
+// with the name constructed from the VM namespace, the VM name, and the volume
+// usage type.
+// At the same time, we expect one lease to be there with similar naming
+// structure except the volume usage type we have one lease for all resources
+// in containerd.
+//
+// == Chapter III: Kernel
+//
+// We do the same checks for kernel. The kernel and volume works the same way,
+// and the real reason why we test it here because the usage type is different,
+// and we should be able to pull and mount them.
+//
+// == Chapter IV: Delete
+//
+// We should be able to delete both the volume and the kernel image. After we
+// delete them, we should see zero images in containerd.
+func TestImageService_Integration(t *testing.T) {
+	if !runContainerDTests() {
+		t.Skip("skipping containerd image service integration test")
+	}
+
+	RegisterTestingT(t)
+
+	//
+	// Preface
+	//
+
+	client, ctx := testCreateClient(t)
+	namespaceCtx := namespaces.WithNamespace(ctx, testContainerdNS)
+
+	imageSvc := containerd.NewImageServiceWithClient(&containerd.Config{
+		SnapshotterKernel: testSnapshotter,
+		SnapshotterVolume: testSnapshotter,
+		Namespace:         testContainerdNS,
+	}, client)
+
+	inputGetAndMount := &ports.ImageMountSpec{
+		ImageName:    getTestVolumeImage(),
+		Owner:        fmt.Sprintf("%s/%s", testOwnerNamespace, testOwnerName),
+		OwnerUsageID: testOwnerUsageID,
+		Use:          models.ImageUseVolume,
+	}
+	inputGet := &ports.ImageSpec{
+		ImageName: inputGetAndMount.ImageName,
+		Owner:     inputGetAndMount.Owner,
+	}
+	expectedSnapshotName := fmt.Sprintf(
+		"flintlock/%s/%s/%s",
+		testOwnerNamespace,
+		testOwnerName,
+		testOwnerUsageID,
+	)
+	expectedLeaseName := fmt.Sprintf("flintlock/%s/%s", testOwnerNamespace, testOwnerName)
+
+	defer func() {
+		// Make sure it's deleted.
+		client.ImageService().Delete(namespaceCtx, getTestKernelImage())
+		client.ImageService().Delete(namespaceCtx, getTestVolumeImage())
+		client.SnapshotService(testSnapshotter).Remove(namespaceCtx, expectedSnapshotName)
+		leases, _ := client.LeasesService().List(namespaceCtx)
+		for _, lease := range leases {
+			client.LeasesService().Delete(namespaceCtx, lease)
```

Comment:
i am not sure this is working. i have a lease left over after running

when i wrap with 
```suggestion
			Expect(client.LeasesService().Delete(namespaceCtx, lease)).To(Succeed())
```
it does not error so 🤷‍♀️ 

(actually might be good to do Succeed check like that just so we don't get silent failures)

Reply:
In the defer cleanup section I would never put test. That's there to ignore everything just try to delete. We have delete test at the end of the test function. This defer call is there in case the test was interrupted/cancelled/killed/whatevercanhappen.

Reply:
it doesn;t need to be a "test" test, just so that if things are left around because deletes fail we actually know about it, rather than being surprised when the next test run hits a conflict

Reply:
The issue with `Expect` it stops when it encounters an error. That's why I added the defer block there. if the test fails anywhere, the rest will not be called, and in this case the cleanup does it without any error handling, just do it.

Reply:
So the only reasonable addition to this would be
```
if err := client.LeasesService().Delete(namespaceCtx, lease); err != nil {
  // log the error.
}
```
I can add some logging in cleanup.

Reply:
aaah i see, yeh good point